extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -2198,4 +2219,100 @@ Runnable updateCurrentAndLatestOffsets()
       }
     };
   }
+
+  /**
+   * Collect row ingestion stats from all tasks managed by this supervisor.
+   *
+   * @return A map of groupId->taskId->task row stats
+   *
+   * @throws InterruptedException
+   * @throws ExecutionException
+   * @throws TimeoutException
+   */
+  private Map<String, Object> getCurrentTotalStats() throws InterruptedException, ExecutionException, TimeoutException
+  {
+    Map<String, Object> allStats = Maps.newHashMap();
+    final List<ListenableFuture<StatsFromTaskResult>> futures = new ArrayList<>();
+
+    for (int groupId : taskGroups.keySet()) {
+      TaskGroup group = taskGroups.get(groupId);
+      for (String taskId : group.taskIds()) {
+        futures.add(
+            Futures.transform(
+                taskClient.getMovingAveragesAsync(taskId),
+                (Function<Map<String, Object>, StatsFromTaskResult>) (currentStats) -> {
+                  return new StatsFromTaskResult(
+                      groupId,
+                      taskId,
+                      currentStats
+                  );
+                }
+            )
+        );
+      }
+    }
+
+    for (int groupId : pendingCompletionTaskGroups.keySet()) {
+      TaskGroup group = taskGroups.get(groupId);
+      for (String taskId : group.taskIds()) {
+        futures.add(
+            Futures.transform(
+                taskClient.getMovingAveragesAsync(taskId),
+                (Function<Map<String, Object>, StatsFromTaskResult>) (currentStats) -> {
+                  return new StatsFromTaskResult(
+                      groupId,
+                      taskId,
+                      currentStats
+                  );
+                }
+            )
+        );
+      }
+    }
+
+    List<StatsFromTaskResult> results = Futures.successfulAsList(futures).get(futureTimeoutInSeconds, TimeUnit.SECONDS);
```

Comment:
Not sure about considering only successful futures. Do we need to check the number of successful futures is same with the number of running tasks?

Reply:
I think it makes sense to treat this as a "best effort" type call, so if some tasks don't return results, we should still return the stats for the ones that succeeded.

I added a check so that the number of results is the same as the number of futures (though these should always be the same based on Futures.successfulAsList docs).

I also added an error log that prints the groupId + taskId for futures that didn't succeed.

Reply:
Thanks. 

> I added a check so that the number of results is the same as the number of futures (though these should always be the same based on Futures.successfulAsList docs).

Yeah, [this check](https://github.com/druid-io/druid/pull/5748/files/2d954d7056a1476f55d8535b981597934188bca5..2fe8122ac23e031c51c8a3541e7c044dcb505441#diff-6eee87b3aa4eb3a516965fe6e93e25a4R2277) looks to have no point. Please remove it.

Reply:
removed check